### PR TITLE
Add/update ceph csi snapshotter rbac and sidecar deployments

### PR DESCRIPTION
**Pull request by humblec**
**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

[test full]
[test ceph]

### CODE REVIEW

File: cluster/charts/rook-ceph/templates/clusterrole.yaml [hunk 1]
```diff
@@ -296,6 +296,24 @@ rules:
   - apiGroups: [""]
     resources: ["persistentvolumeclaims/status"]
     verbs: ["update", "patch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
```
Comment (Madhu-1) [inline cluster/charts/rook-ceph/templates/clusterrole.yaml:299]: need same changes in common.yaml and e2e manifests also

**Comment by Madhu-1** [comment]
added skip CI  to avoid running E2E till we get ceph-csi release out

File: cluster/charts/rook-ceph/templates/clusterrole.yaml (hunk 1 above)
Comment (humblec) [reply cluster/charts/rook-ceph/templates/clusterrole.yaml:299]: was already on it, common is already addressed. Looking into e2e.

**Comment by humblec** [comment]
> a couple of things needed.
> 
>     * documentation for cephfs snapshot, restore and clone
> 
>     * csi  image version change
> 
> 
> This PR should not be merged till v.3.1.0 is released

Yeah.. :+1:

File: cluster/charts/rook-ceph/templates/clusterrole.yaml (hunk 1 above)
Comment (humblec) [reply cluster/charts/rook-ceph/templates/clusterrole.yaml:299]: done. thanks .

**Comment by humblec** [comment]
I will push a documentation PR as a followup.

File: cluster/examples/kubernetes/ceph/common.yaml [hunk 2]
```diff
@@ -1582,6 +1582,24 @@ rules:
   - apiGroups: [""]
     resources: ["events"]
     verbs: ["list", "watch", "create", "update", "patch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
```
Comment (Madhu-1) [inline cluster/examples/kubernetes/ceph/common.yaml:1585]: These new rbac.need to be added for upgrade yaml file

Comment (travisn) [reply cluster/examples/kubernetes/ceph/common.yaml:1585]: Yes, should be good in upgrade-from-v1.3-apply.yaml.
Note that it will not be applied for those who have already upgraded to v1.4.0. This should be ok if we just put a note in the release notes that this must be applied if you want to use the new cephfs snapshotting feature.

File: cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml [hunk 3]
```diff
@@ -34,6 +34,21 @@ spec:
           volumeMounts:
             - name: socket-dir
               mountPath: /csi
+        - name: csi-snapshotter
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml:37]: I believe we need to conditionally enable the snapshotter if the k8s version supports it, otherwise the container logs for the snapshotter will spew errors. We also have this issue for the rbd snapshotter. However, I don't see an issue opened for it yet.

File: cluster/examples/kubernetes/ceph/common.yaml (hunk 2 above)
Comment (humblec) [reply cluster/examples/kubernetes/ceph/common.yaml:1585]: Done. added to `1.3` upgrade yaml. Sure, after 1.4.0 they could apply the same and get it working.

Comment (humblec) [reply cluster/examples/kubernetes/ceph/common.yaml:1585]: @travisn @Madhu-1 ptal.

File: cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml (hunk 3 above)
Comment (humblec) [reply cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml:37]: Yeah. that would be good and can both ( RBD and CephFS ) handled together in a followup PR :+1:

**Comment by humblec** [comment]
@travisn @Madhu-1  I have added the documentation too in this PR itself. PTAL .

File: Documentation/ceph-csi-snapshot.md [hunk 4]
```diff
@@ -99,3 +100,78 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
...
 kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/snapshotclass.yaml
 ```
+
+## CephFS Snapshots
+
+### SnapshotClass
+
+In [snapshotClass](https://github.com/rook/rook/tree/{{ branchName }}/cluster/examples/kubernetes/ceph/csi/cephfs/snapshotclass.yaml),
+the `csi.storage.k8s.io/snapshotter-secret-name` parameter should reference the
+name of the secret created for the cephfsplugin and `pool` to reflect the Ceph pool name.
```
Comment (Madhu-1) [inline Documentation/ceph-csi-snapshot.md:110]: why we need pool in cephfs snapshot class?

File: Documentation/ceph-csi-snapshot.md [hunk 5]
```diff
@@ -99,3 +100,78 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
...
+
+Update the value of the `clusterID` field to match the namespace that Rook is
```
Comment (Madhu-1) [inline Documentation/ceph-csi-snapshot.md:112]: this can be explained only once we can move this to the top to keep common for both rbd and cephfs

File: Documentation/ceph-csi-snapshot.md [hunk 6]
```diff
@@ -99,3 +100,78 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
...
+running in. When Ceph CSI is deployed by Rook, the operator will automatically
+maintain a configmap whose contents will match this key. By default this is
+"rook-ceph".
+
+```console
+kubectl create -f cluster/examples/kubernetes/ceph/csi/rbd/snapshotclass.yaml
```
Comment (Madhu-1) [inline Documentation/ceph-csi-snapshot.md:118]: why we are creating rbd snapshotclass for cephfs?

File: Documentation/ceph-csi-snapshot.md [hunk 7]
```diff
@@ -99,3 +100,78 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
...
+```
+
+## CephFS snapshot resource Cleanup
+
+To clean your cluster of the resources created by this example, run the following:
+
+```console
+kubectl delete -f cluster/examples/kubernetes/ceph/csi/cephfs/pvc-restore.yaml
+kubectl delete -f cluster/examples/kubernetes/ceph/csi/cephfs/snapshot.yaml
+kubectl delete -f cluster/examples/kubernetes/ceph/csi/cephfs/snapshotclass.yaml
```
Comment (Madhu-1) [inline Documentation/ceph-csi-snapshot.md:176]: please add the limitations for cephfs snapshots to this doc so that people will be aware of it
* Limit of 400 snapshots per filesystem
* PVC cannot be deleted if it's having snapshots

File: cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml (hunk 3 above)
Comment (Madhu-1) [reply cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml:37]: is it possible to do it before the release?

File: tests/framework/installer/ceph_csi_templates.go [hunk 8]
```diff
@@ -332,6 +332,21 @@ const (
               volumeMounts:
                 - name: socket-dir
                   mountPath: /csi
+            - name: csi-snapshotter
+              image:  {{ .SnapshotterImage }}
+              args:
+                - "--csi-address=$(ADDRESS)"
+                - "--v=5"
+                - "--timeout=60s"
```
Comment (Madhu-1) [inline tests/framework/installer/ceph_csi_templates.go:340]: why the 60s? leader election flags are missing here

File: pkg/operator/ceph/csi/spec.go [hunk 9]
```diff
@@ -96,7 +96,7 @@ var (
 // manually challenging.
 var (
 	// image names
-	DefaultCSIPluginImage   = "quay.io/cephcsi/cephcsi:v3.0.0"
+	DefaultCSIPluginImage   = "quay.io/cephcsi/cephcsi:v3.1.0"
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/spec.go:99]: need to add same to https://github.com/rook/rook/blob/master/pkg/operator/ceph/csi/version.go#L32

Comment (Madhu-1) [inline pkg/operator/ceph/csi/spec.go:99]: same change need to be done for helm charts, operator.yaml openshift-operator.yaml and also need to updated the upgrade-documentation also

File: tests/framework/installer/ceph_csi_templates.go (hunk 8 above)
Comment (humblec) [reply tests/framework/installer/ceph_csi_templates.go:341]: Increased to 150. also added leader election for both cephfs and RBD as it was missing. :+1:

**Comment by humblec** [comment]
@Madhu-1 comments are addressed. Flag based enablement can be done later or for next patch release of rook.

File: cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml (hunk 3 above)
Comment (humblec) [reply cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml:37]: we can defer it from this release and get it done for next patch release.

File: Documentation/ceph-csi-snapshot.md [hunk 10]
```diff
@@ -4,23 +4,19 @@ weight: 3250
...
 
 Note: If the Kubernetes distributor you are using does not supports the snapshot beta,
 still you can use the Alpha snapshots. refer to
 [snapshot](https://github.com/rook/rook/blob/release-1.3/Documentation/ceph-csi-drivers.md#rbd-snapshots)
 on how to use snapshots.
 
-### SnapshotClass
-
-You need to create the `SnapshotClass`. The purpose of a `SnapshotClass` is
+- Create the `SnapshotClass`. The purpose of a `SnapshotClass` is
```
Comment (Madhu-1) [inline Documentation/ceph-csi-snapshot.md:19]: the snapshot class cannot be `Prerequisites` isn't it? as we are talking about how to create snapshotclass below.

File: Documentation/ceph-csi-snapshot.md [hunk 11]
```diff
@@ -99,3 +105,77 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
 kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/snapshot.yaml
 kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/snapshotclass.yaml
 ```
+
+## CephFS Snapshots
+
+NOTE: At present, there is a limit of 400 snapshots per cephFS filesystem. Also PVC cannot be deleted if it's having snapshots.
```
Comment (Madhu-1) [inline Documentation/ceph-csi-snapshot.md:111]: can we highlight this one?

File: cluster/charts/rook-ceph/values.yaml [hunk 12]
```diff
@@ -239,7 +239,7 @@ csi:
   #rbdLivenessMetricsPort: 9080
   #kubeletDirPath: /var/lib/kubelet
   #cephcsi:
-    #image: quay.io/cephcsi/cephcsi:v3.0.0
+    #image: quay.io/cephcsi/cephcsi:v3.1.0
```
Comment (Madhu-1) [inline cluster/charts/rook-ceph/values.yaml:242]: the same change is also required in operator.yaml and openshift-operator.yaml

File: tests/framework/installer/ceph_csi_templates.go [hunk 13]
```diff
@@ -332,6 +333,22 @@ const (
...
```
Comment (Madhu-1) [inline tests/framework/installer/ceph_csi_templates.go:341]: still, I see 60sec here

File: Documentation/ceph-csi-snapshot.md (hunk 11 above)
Comment (humblec) [reply Documentation/ceph-csi-snapshot.md]: sure. done

File: Documentation/ceph-csi-snapshot.md (hunk 10 above)
Comment (humblec) [reply Documentation/ceph-csi-snapshot.md]: Adjusted this a bit.. Ptal.

File: Documentation/ceph-csi-snapshot.md [hunk 14]
```diff
@@ -99,3 +101,82 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
 kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/snapshot.yaml
 kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/snapshotclass.yaml
 ```
+
+## CephFS Snapshots
+
+**NOTE: At present, there is a limit of 400 snapshots per cephFS filesystem. Also PVC cannot be deleted if it's having snapshots.
+make sure the snapshots on the PVC are deleted before you delete the PVC.**
```
Comment (Madhu-1) [inline Documentation/ceph-csi-snapshot.md:108]:
```suggestion
Make sure all the snapshots on the PVC are deleted before you delete the PVC.**
```

File: Documentation/ceph-csi-snapshot.md [hunk 15]
```diff
@@ -4,23 +4,19 @@ weight: 3250
...
 
 Note: If the Kubernetes distributor you are using does not supports the snapshot beta,
 still you can use the Alpha snapshots. refer to
 [snapshot](https://github.com/rook/rook/blob/release-1.3/Documentation/ceph-csi-drivers.md#rbd-snapshots)
 on how to use snapshots.
 
-### SnapshotClass
-
-You need to create the `SnapshotClass`. The purpose of a `SnapshotClass` is
+- We also need a `SnapshotClass` for volume snapshot to work. The purpose of a `SnapshotClass` is
```
Comment (Madhu-1) [inline Documentation/ceph-csi-snapshot.md:19]: is this a prereq?

File: Documentation/ceph-csi-snapshot.md (hunk 14 above)
Comment (humblec) [reply Documentation/ceph-csi-snapshot.md]: sure.. done

File: Documentation/ceph-csi-snapshot.md (hunk 15 above)
Comment (humblec) [reply Documentation/ceph-csi-snapshot.md:19]: Should be fine. As snapshot can not be created without SnapshotClass and also as we have more details in the coming sections.

File: Documentation/ceph-csi-snapshot.md [hunk 16]
```diff
@@ -4,23 +4,19 @@ weight: 3250
 indent: true
 ---
 
-## RBD Snapshots
-
-### Prerequisites
+## Prerequisites
 
-1. Requires Kubernetes v1.17+ which supports snapshot beta.
+- Requires Kubernetes v1.17+ which supports snapshot beta.
```
Comment (travisn) [inline Documentation/ceph-csi-snapshot.md:9]: nit: `*` is used in rook instead of `-`
```suggestion
* Requires Kubernetes v1.17+ which supports snapshot beta.
```

File: Documentation/ceph-csi-snapshot.md [hunk 17]
```diff
@@ -99,3 +101,82 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
 kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/snapshot.yaml
 kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/snapshotclass.yaml
 ```
+
+## CephFS Snapshots
+
+**NOTE: At present, there is a limit of 400 snapshots per cephFS filesystem. Also PVC cannot be deleted if it's having snapshots.
```
Comment (travisn) [inline Documentation/ceph-csi-snapshot.md:107]: How about moving this to a "limitations" section later in the doc and put it as bullet points? These don't seem like blocking issues that are important at the beginning of the section.
```suggestion
**NOTE: There is a limit of 400 snapshots per CephFS filesystem. Also the PVC cannot be deleted if it has snapshots.
```

File: Documentation/ceph-csi-snapshot.md [hunk 18]
```diff
@@ -99,3 +101,82 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
...
+```console
+kubectl create -f cluster/examples/kubernetes/ceph/csi/cephfs/snapshotclass.yaml
+```
+
+### Volumesnapshot
+
+In [snapshot](https://github.com/rook/rook/tree/{{ branchName }}/cluster/examples/kubernetes/ceph/csi/cephfs/snapshot.yaml),
+`volumeSnapshotClassName` should be the name of the `VolumeSnapshotClass`
+previously created. The `persistentVolumeClaimName` should be the name of the
+PVC which is already created by CephFS CSI driver.
```
Comment (travisn) [inline Documentation/ceph-csi-snapshot.md:131]:
```suggestion
PVC which is already created by the CephFS CSI driver.
```

File: Documentation/ceph-csi-snapshot.md [hunk 19]
```diff
@@ -99,3 +101,82 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
...
+kubectl get volumesnapshotclass
+NAME                        DRIVER                          DELETIONPOLICY   AGE
+csi-cephfslugin-snapclass   rook-ceph.cephfs.csi.ceph.com   Delete           3h55m
+
+kubectl get volumesnapshot
+NAME                  READYTOUSE   SOURCEPVC   SOURCESNAPSHOTCONTENT  RESTORESIZE   SNAPSHOTCLASS                SNAPSHOTCONTENT                                   CREATIONTIME   AGE
+cephfs-pvc-snapshot   true         cephfs-pvc                         1Gi           csi-cephfsplugin-snapclass   snapcontent-34476204-a14a-4d59-bfbc-2bbba695652c  3h50m          3h51m
+```
+
+The snapshot will be ready to restore to a new PVC when `READYTOUSE` field of the
```
Comment (travisn) [inline Documentation/ceph-csi-snapshot.md:149]:
```suggestion
The snapshot will be ready to restore to a new PVC when the `READYTOUSE` field of the
```

File: Documentation/ceph-csi-snapshot.md [hunk 20]
```diff
@@ -99,3 +101,82 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
...
+**NOTE: At present, there is a limit of 400 snapshots per cephFS filesystem. Also PVC cannot be deleted if it's having snapshots.
+make sure all the snapshots on the PVC are deleted before you delete the PVC.**
+
+### SnapshotClass
+
+In [snapshotClass](https://github.com/rook/rook/tree/{{ branchName }}/cluster/examples/kubernetes/ceph/csi/cephfs/snapshotclass.yaml),
+the `csi.storage.k8s.io/snapshotter-secret-name` parameter should reference the
+name of the secret created for the cephfsplugin.
+
+In the snapshotclass, update the value of the `clusterID` field to match the namespace that Rook is
```
Comment (travisn) [inline Documentation/ceph-csi-snapshot.md:116]: When mentioning the properties, how about adding it as a bullet list and start each bullet with the setting name? In the paragraph form it is easy to overlook the setting names. This suggestion applies to all the sections where you have new settings to document. I guess this also affects the RBD docs, so this could be a separate PR.

File: cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml (hunk 3 above)
Comment (travisn) [reply cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml:37]: Can we get this early next week?

File: Documentation/ceph-csi-snapshot.md (hunk 20 above)
Comment (humblec) [reply Documentation/ceph-csi-snapshot.md:116]: Sure.. we could do a follow-up PR for this.

**Comment by humblec** [comment]
@travisn addressed the comments and also squashed 3/4 commits. PTAL.  The Ceph CSI v3.1.0 is also available now https://github.com/ceph/ceph-csi/releases/tag/v3.1.0 :+1:

File: cluster/charts/rook-ceph/templates/clusterrole.yaml (hunk 1 above)
Comment (travisn) [inline cluster/charts/rook-ceph/templates/clusterrole.yaml:299]: @humblec @Madhu-1 In the rook patch releases, our goal is to make RBAC changes optional. What will happen if this RBAC is not applied? Will the snapshotter container just write errors to the log? Or will there be any other side effect?

Comment (Madhu-1) [reply cluster/charts/rook-ceph/templates/clusterrole.yaml:299]: User cannot create and delete snapshots

Comment (Madhu-1) [reply cluster/charts/rook-ceph/templates/clusterrole.yaml:299]: Yes snapshotter will keep writing error to logs
